packages/json-api/src/Registry.ts

```diff
@@ -0,0 +1,92 @@
+import { JSONValue, JSONAPIDocument, JSONAPIResource } from "./jsonApiTypes";
+import {
+  IncludesMap,
+  MapObject,
+  Registry,
+  Resource,
+  ResourceParseOptions,
+  ResourceSpec
+} from "./types";
+import { identity, kebabCaseDeep, camelCaseDeep, TransformFunc } from "./utils";
+import ResourceImpl from "./Resource";
+
+export type RegistryCreateOptions = {
+  keyTransform?: "kebab";
+};
+
+export default class RegistryImpl implements Registry {
+  private readonly options?: RegistryCreateOptions;
+
+  private readonly resources: MapObject<Resource>;
+
+  public readonly keyTransformFunc: TransformFunc;
+
+  public readonly keyParseFunc: TransformFunc;
+
+  constructor(options?: RegistryCreateOptions) {
+    this.options = options;
+    this.resources = {};
+
+    this.keyTransformFunc =
+      this.options?.keyTransform === "kebab" ? kebabCaseDeep : identity;
+    this.keyParseFunc =
+      this.options?.keyTransform === "kebab" ? camelCaseDeep : identity;
+  }
+
+  define(type: string, spec: ResourceSpec): Resource {
+    const resource = new ResourceImpl(this, type, spec);
+    if (type in this.resources) {
+      throw new Error(`Resource ${type} is already defined`);
+    }
+    this.resources[type] = resource;
+    return resource;
+  }
+
+  find(type: string): Resource | null {
+    return this.resources[type];
+  }
+
+  format(type: string, data: JSONValue): JSONAPIDocument {
+    const resource = this.find(type);
+    if (!resource) {
+      throw new Error(`Cannot find resource ${type}`);
+    }
+    return resource.document(data);
+  }
+
+  parse(document: JSONAPIDocument, options?: ResourceParseOptions): JSONValue {
+    const parseResource = (
+      data: JSONAPIResource,
+      includesMap?: IncludesMap
+    ) => {
+      const resource = this.find(data.type);
+
+      if (!resource) {
+        throw new Error(`Cannot find resource ${data.type}`);
+      }
+
+      return resource.parse(data, includesMap, options);
+    };
+
+    const includesMap: IncludesMap | undefined = Array.isArray(
+      document.included
+    )
+      ? document.included.reduce((accum: IncludesMap, res) => {
+          const resource = parseResource(res);
+          return {
+            ...accum,
+            [`${res.type}:${resource.id as string}`]: resource
+          };
+        }, {})
+      : undefined;
+
+    if (Array.isArray(document.data)) {
+      return (document.data as JSONAPIResource[]).map(elem =>
+        parseResource(elem, includesMap)
+      );
+    }
+    return parseResource(document.data as JSONAPIResource, includesMap);
+  }
+}
```

Comment:
After moving throwing error we can remove `parseResource`

```suggestion
  parse(document: JSONAPIDocument, options?: ResourceParseOptions): JSONValue {
    const includesMap: IncludesMap | undefined = Array.isArray(
      document.included
    )
      ? document.included.reduce((accum: IncludesMap, res) => ({
            ...accum,
            [`${res.type}:${resource.id as string}`]: this.find(res.type).parse(res),
          }), {})
      : undefined;

    if (Array.isArray(document.data)) {
      return (document.data as JSONAPIResource[]).map(elem =>
        this.find(res.type).parse(elem, includesMap)
      );
    }
    return this.find(res.type).parse(document.data as JSONAPIResource, includesMap);
  }
}
```

Reply:
See my comments above - at the moment I really prefer to leave API unchanged.